src/librbd/Utils.cc

```diff
@@ -242,5 +247,152 @@ int get_config_key(librados::Rados& rados, const std::string& uri,
+    lderr(cct) << "could not read ceph conf for " << description << ": "
+	             << cpp_strerror(r) << dendl;
+  }
+
+  // preserve cluster-specific config settings before applying environment/cli
+  // overrides
+  std::map<std::string, std::string> config_values;
+  if (strip_cluster_overrides) {
+    // remote peer connections shouldn't apply cluster-specific
+    // configuration settings
+    for (auto& key : UNIQUE_PEER_CONFIG_KEYS) {
+      config_values[key] = new_cct->_conf.get_val<std::string>(key);
+    }
+  }
+
+  new_cct->_conf.parse_env(new_cct->get_module_type());
+
+  // librados::Rados::conf_parse_env
+  std::vector<const char*> env_args;
+  r = new_cct->_conf.parse_argv(env_args);
+  if (r < 0) {
+    lderr(cct) << "could not parse environment for " << description << ":"
+               << cpp_strerror(r) << dendl;
+    new_cct->put();
+    return r;
+  }
+  new_cct->_conf.parse_env(new_cct->get_module_type());
+
+  if (!args.empty()) {
+    // librados::Rados::conf_parse_argv
+    r = new_cct->_conf.parse_argv(args);
+    if (r < 0) {
+      lderr(cct) << "could not parse command line args for " << description
+                 << ": " << cpp_strerror(r) << dendl;
+      new_cct->put();
+      return r;
+    }
+  }
+
+  if (strip_cluster_overrides) {
+    // remote peer connections shouldn't apply cluster-specific
+    // configuration settings
+    for (auto& pair : config_values) {
+      auto value = new_cct->_conf.get_val<std::string>(pair.first);
+      if (pair.second != value) {
+        ldout(cct, 20) << "reverting global config option override: "
+                       << pair.first << ": " << value << " -> " << pair.second
+                       << dendl;
+        r = new_cct->_conf.set_val(pair.first, pair.second);
+        if (r < 0) {
+          lderr(cct) << "failed to set " << pair.first << " config for "
+                     << description << ": " << cpp_strerror(r) << dendl;
+          new_cct->put();
+          return r;
+        }
+      }
+    }
+  }
+
+  if (!admin_socket.empty()) {
+    r = new_cct->_conf.set_val("admin_socket", admin_socket);
+    if (r < 0) {
+      lderr(cct) << "failed to set admin_socket config for " << description
+                 << ": " << cpp_strerror(r) << dendl;
+      new_cct->put();
+      return r;
+    }
+  }
+
+  if (!mon_host.empty()) {
+    r = new_cct->_conf.set_val("mon_host", mon_host);
+    if (r < 0) {
+      lderr(cct) << "failed to set mon_host config for " << description << ": "
+                 << cpp_strerror(r) << dendl;
+      new_cct->put();
+      return r;
+    }
+  }
+
+  if (!key.empty()) {
+    r = new_cct->_conf.set_val("key", key);
+    if (r < 0) {
+      lderr(cct) << "failed to set key config for " << description << ": "
+                 << cpp_strerror(r) << dendl;
+      new_cct->put();
+      return r;
+    }
+  }
+
+  // disable unnecessary librbd cache
+  r = new_cct->_conf.set_val("rbd_cache", "false");
```

Comment:
nit: I am not sure we want this in generic case. It is useful to disable for rbd-mirror. Probably it is also not very useful for external migration, still in general case...

So probably it would be cleaner to do this in rbd-mirror, after `init_rados`, but I would not mind any way.